custom/icds/location_rationalization/models.py

```diff
@@ -0,0 +1,153 @@
+from abc import ABCMeta, abstractmethod
+from datetime import datetime
+
+from django.db import transaction
+
+from custom.icds.location_rationalization.const import (
+    DEPRECATED_AT,
+    DEPRECATED_TO,
+    DEPRECATES,
+    DEPRECATION_OPERATION,
+    EXTRACT_OPERATION,
+    MERGE_OPERATION,
+    MOVE_OPERATION,
+    SPLIT_OPERATION,
+)
+
+
+class Transition(object):
+    def __init__(self, operation, old_locations, new_locations):
+        self.operation = {
+            MERGE_OPERATION: MergeOperation,
+            SPLIT_OPERATION: SplitOperation,
+            EXTRACT_OPERATION: ExtractOperation,
+            MOVE_OPERATION: MoveOperation
+        }[operation](old_locations, new_locations)
+
+    def valid(self):
+        return self.operation.valid()
+
+    def perform(self):
+        return self.operation.perform()
+
+
+class BaseOperation(metaclass=ABCMeta):
+    type = None
+
+    def __init__(self, old_locations, new_locations):
+        self.old_locations = old_locations
+        self.new_locations = new_locations
+
+    def valid(self):
+        """
+        Invalid if
+        1. there are no locations
+        2. if any of the old locations have already been deprecated
+        3. if any of the new locations has already been a part of a deprecation
+        :return:
+        """
+        if not self.old_locations or not self.new_locations:
+            return False
+        for old_location in self.old_locations:
+            if (DEPRECATED_TO in old_location.metadata or
+                    DEPRECATED_AT in old_location.metadata or
+                    DEPRECATION_OPERATION in old_location.metadata):
+                return False
+        for new_location in self.new_locations:
+            if DEPRECATES in new_location.metadata:
+                return False
+
+    @abstractmethod
+    def perform(self):
+        raise NotImplementedError
+
+
+class MergeOperation(BaseOperation):
+    type = MERGE_OPERATION
+
+    def valid(self):
+        valid = super().valid()
+        if not valid:
+            return False
+        return len(self.old_locations) > 1 and len(self.new_locations) == 1
+
+    def perform(self):
+        with transaction.atomic():
+            new_location = self.new_locations[1]
```

Comment:
if `len(self.new_locations) == 1` shouldn't this be `self.new_locations[0]`?

Reply:
Yes sorry.

Reply:
https://github.com/dimagi/commcare-hq/pull/26898/commits/e22af0a6af64169e77cb231e2ac7f5b90bd7eeef